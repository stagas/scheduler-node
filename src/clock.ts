export class Clock {
  buffer = new Float64Array(
    new SharedArrayBuffer(3 * Float64Array.BYTES_PER_ELEMENT)
  )

  get coeff() {
    return this.buffer[0]
  }
  set coeff(value: number) {
    this.buffer[0] = value
  }

  get internalTime() {
    return this.buffer[1]
  }
  set internalTime(value: number) {
    this.buffer[1] = value
  }

  get offsetFrame() {
    return this.buffer[2]
  }
  set offsetFrame(value: number) {
    this.buffer[2] = value
  }
}
