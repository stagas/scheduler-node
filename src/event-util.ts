import { createMidiNoteEvents } from 'webaudio-tools/midi-events'

export function getMidiEventsForNotes(notes: NoteEvent[], bars?: number, sampleRate: number = 44100) {
  const midiEvents = []

  // eslint-disable-next-line prefer-const
  for (let [i, [time, note, velocity, length]] of notes.entries()) {
    if (time != null && note != null && velocity != null) {
      if (!length) {
        length = ((notes[i + 1]?.[0] ?? bars) - time) - (1 / sampleRate) * 2
      }
      midiEvents.push(...createMidiNoteEvents(time, note, velocity, length))
    }
  }

  return midiEvents
}

export type NoteEvent = [number, number, number] | [number, number, number, number]
