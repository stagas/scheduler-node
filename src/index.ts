export * from './scheduler-event'
export * from './scheduler-node'
export * from './scheduler-target-node'
export * from './clock'
