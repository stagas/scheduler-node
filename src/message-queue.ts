export class MessageQueue {
  buffer: Float64Array | Float32Array = new Float64Array(
    new SharedArrayBuffer(8192 * 2 * Float64Array.BYTES_PER_ELEMENT)
  )

  constructor(messageQueue: Partial<MessageQueue> = {}) {
    Object.assign(this, messageQueue)
  }

  toJSON() {
    return {
      buffer: this.buffer,
    }
  }

  get readPtr() {
    return this.buffer[0]
  }

  set readPtr(index) {
    this.buffer[0] = index
  }

  get writePtr() {
    return this.buffer[1]
  }

  set writePtr(index) {
    this.buffer[1] = index
  }

  clear() {
    this.buffer.fill(-Infinity)
    this.readPtr = 2
    this.writePtr = 2
    return this
  }

  push(...values: number[]) {
    const buffer = this.buffer
    const ptr = this.writePtr

    buffer.set(values, ptr)

    const nextPtr = ptr + values.length
    if (nextPtr < this.buffer.length) {
      this.writePtr = nextPtr
    }
  }

  shift() {
    const buffer = this.buffer
    const ptr = this.readPtr

    const value = buffer[ptr]
    if (!isFinite(value)) {
      this.readPtr = this.writePtr = 2
      return undefined
    }

    buffer[ptr] = -Infinity

    this.readPtr = ptr + 1
    return value
  }

  slice(length: number) {
    const buffer = this.buffer
    const ptr = this.readPtr

    const slice = buffer.subarray(ptr, ptr + length)
    if (!isFinite(slice[0])) {
      this.buffer.fill(-Infinity, 2, 2 + ptr)
      this.readPtr = this.writePtr = 2
      return undefined
    }

    this.readPtr = ptr + length
    return slice
  }
}
