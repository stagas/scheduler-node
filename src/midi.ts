export class MIDIMessageEvent extends Event implements WebMidi.MIDIMessageEvent {
  data: Uint8Array
  receivedTime!: number
  receivedFrame!: number
  offsetFrame!: number
  deltaFrame!: number
  constructor(type: string, options: any) {
    super(type)
    this.data = options.data
  }
}
