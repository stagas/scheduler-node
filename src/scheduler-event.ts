import { cheapRandomId } from 'everyday-utils'
import { ImmSet } from 'immutable-map-set'
import { createMidiNoteEvents } from 'webaudio-tools/midi-events'

import { MessageQueue } from './message-queue'
import { SchedulerNode } from './scheduler-node'
import { SchedulerTargetNode } from './scheduler-target-node'

export class SchedulerEvent {
  id = cheapRandomId()

  midiEvent = { data: new Uint8Array(3), receivedTime: 0 }

  constructor(data: Partial<SchedulerEvent> = {}) {
    this.id = data.id ?? this.id
    if (data.midiEvent) {
      this.midiEvent.data = data.midiEvent.data
      this.midiEvent.receivedTime = data.midiEvent.receivedTime
    }
  }

  toJSON() {
    return {
      id: this.id,
      midiEvent: {
        receivedTime: this.midiEvent.receivedTime ?? 0,
        data: this.midiEvent.data,
      },
    }
  }
}

export class SchedulerTarget {
  id = cheapRandomId()

  midiQueue = new MessageQueue().clear()

  constructor(data: Partial<SchedulerTarget> = {}) {
    Object.assign(this, data)
  }
}

export class SchedulerEventGroup {
  id = cheapRandomId()

  targets = new ImmSet<SchedulerTarget>()
  events = new ImmSet<SchedulerEvent>()

  loopPoints = new Float64Array(
    new SharedArrayBuffer(
      3 // [loopActive, loopStart, loopEnd]
      * Float64Array.BYTES_PER_ELEMENT
    )
  )

  constructor(eventGroup: Partial<SchedulerEventGroup> = {}) {
    Object.assign(this, eventGroup)
  }

  toJSON() {
    return {
      id: this.id,
      targets: this.targets,
      events: this.events,
      loopPoints: this.loopPoints,
    }
  }

  get loop() {
    return !!this.loopPoints[0]
  }
  set loop(value: boolean) {
    this.loopPoints[0] = +value
  }

  get loopStart() {
    return this.loopPoints[1]
  }
  set loopStart(seconds: number) {
    this.loopPoints[1] = seconds
  }

  get loopEnd() {
    return this.loopPoints[2]
  }
  set loopEnd(seconds: number) {
    this.loopPoints[2] = seconds
  }

  replaceAllWithNotes(notes: [number, number, number, number][]) {
    const midiEvents = []

    for (const [time, note, velocity, length] of notes) {
      midiEvents.push(...createMidiNoteEvents(time, note, velocity, length))
    }

    this.events = this.events.clear()

    for (const midiEvent of midiEvents) {
      const event = new SchedulerEvent({ midiEvent })
      this.events = this.events.add(event)
    }

    return midiEvents
  }
}

export class SchedulerEventGroupNode extends EventTarget {
  eventGroup = new SchedulerEventGroup()
  targetNodes = new Set<SchedulerTargetNode>()

  declare onconnectchange: (ev: CustomEvent) => void

  constructor(public schedulerNode: SchedulerNode) {
    super()
    this.schedulerNode.eventGroups.add(this.eventGroup)
  }

  destroy() {
    this.schedulerNode.eventGroups.delete(this.eventGroup)
  }

  connect(targetNode: SchedulerTargetNode) {
    this.schedulerNode.connect(targetNode)
    this.targetNodes.add(targetNode)
    this.eventGroup.targets = this.eventGroup.targets.add(targetNode.schedulerTarget)
    this.dispatchEvent(new CustomEvent('connectchange'))
    return targetNode
  }

  disconnect(targetNode: SchedulerTargetNode) {
    this.schedulerNode.disconnect(targetNode)
    this.targetNodes.delete(targetNode)
    this.eventGroup.targets = this.eventGroup.targets.delete(targetNode.schedulerTarget)
    this.dispatchEvent(new CustomEvent('connectchange'))
  }
}
