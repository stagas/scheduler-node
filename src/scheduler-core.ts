import { ImmSet } from 'immutable-map-set'
import { createContext as createObjectifyContext } from 'json-objectify'
import { createContext as createSerializeContext, thru } from 'serialize-whatever'

import { MessageQueue } from './message-queue'
import { SchedulerEvent, SchedulerEventGroup, SchedulerTarget } from './scheduler-event'

const deserializableClasses = [
  SchedulerEventGroup,
  SchedulerEvent,
  SchedulerTarget,
  MessageQueue,
  ImmSet,
]

const { replacer, reviver } = createSerializeContext([
  [Float64Array, [thru, thru]],
  [Uint8Array, [thru, thru]],
])

const { objectify, deobjectify } = createObjectifyContext([Float64Array, Uint8Array])

export const core = {
  pickFromLocal: ['id', 'targets', 'loopPoints'] as (keyof SchedulerEventGroup)[],
  serialize: (data: any) => objectify(data, replacer(data)),
  deserialize: (data: any) => deobjectify(data, reviver(deserializableClasses)) as any,
}
