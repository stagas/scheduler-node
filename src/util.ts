import { SchedulerEvent } from './scheduler-event'

export type EventTuple = [number, SchedulerEvent]

export const getEventsInRange = (
  events: Iterable<SchedulerEvent>,
  loop: boolean,
  loopStart: number,
  loopEnd: number,
  currentTime: number,
  sampleTime: number,
  quantumDurationTime: number,
  playbackStartTime: number,
) => {
  let startTime: number
  let endTime: number
  let offsetTime = 0

  let loopDuration = 0

  const results: EventTuple[] = []

  if (loop) {
    loopDuration = loopEnd - loopStart
    offsetTime = currentTime - playbackStartTime - loopStart
    if (offsetTime < 0) {
      return results
    }

    startTime = offsetTime % loopDuration
    endTime = (startTime + quantumDurationTime) % loopDuration

    if (startTime < sampleTime) startTime = 0
    if (endTime < sampleTime) endTime = 0
  } else {
    startTime = currentTime - playbackStartTime
    endTime = startTime + quantumDurationTime
  }

  for (const event of events) {
    const eventTime = event.midiEvent.receivedTime * 0.001 // ms to seconds
    const isPast = endTime < startTime

    if (
      eventTime === startTime && startTime === endTime
      || (isPast
        ? eventTime >= loopStart && eventTime < endTime
        : eventTime >= startTime && eventTime < endTime)
    ) {
      let receivedTime: number
      if (loop) {
        receivedTime = eventTime
          // ceil or floor whether to go to next bar quantum or this
          + Math[isPast ? 'ceil' : 'floor'](startTime / loopDuration) * loopDuration
          + playbackStartTime
          + (offsetTime - startTime)
      } else {
        // TODO: -startTime ?
        receivedTime = eventTime + playbackStartTime
      }
      receivedTime *= 1000
      results.push([receivedTime, event])
    }
  }

  return results
}
