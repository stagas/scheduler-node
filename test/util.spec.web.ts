import { MIDIMessageEvent } from '../src/midi'
import { SchedulerEvent } from '../src/scheduler-event'
import { getEventsInRange } from '../src/util'

let i = 0
Math.random = () => ++i / 1000

const makeEvent = (x: number, receivedTime: number) => {
  const midiEvent = new MIDIMessageEvent('midimessage', { data: new Uint8Array([x, x, x]) })
  midiEvent.receivedTime = receivedTime * 1000
  return new SchedulerEvent({ midiEvent })
}

describe('getEventsInRange()', () => {
  it('start', () => {
    const events = [
      makeEvent(1, 0),
    ]
    const sampleTime = 1 / 44100
    const results = getEventsInRange(events, true, 0, 1, 0, sampleTime, 128 / 44100, 0)
    expect(results).toMatchSnapshot()
  })

  it('edge', () => {
    const events = [
      makeEvent(1, 0),
    ]
    const quantum = 128 / 44100
    const sample = 1 / 44100
    {
      const results = getEventsInRange(events, true, 0, 1 / 5, 12.8 - quantum, sample, quantum, 0)
      expect(results).toMatchSnapshot()
    }
    {
      const results = getEventsInRange(events, true, 0, 1 / 5, 12.8, sample, quantum, 0)
      expect(results).toMatchSnapshot()
    }
    {
      const results = getEventsInRange(events, true, 0, 1 / 5, 12.8 + quantum, sample, quantum, 0)
      expect(results).toMatchSnapshot()
    }
  })
})
