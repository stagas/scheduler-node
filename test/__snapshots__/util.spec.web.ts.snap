
exports[`getEventsInRange() start 1`] = `
Array [
  Array [
    0,
    Object {
      "id": "255s",
      "midiEvent": Object {
        "data": Uint8Array [
          1,
          1,
          1,
        ],
        "receivedTime": 0,
      },
    },
  ],
]
`;

exports[`getEventsInRange() edge 1`] = `Array []`;

exports[`getEventsInRange() edge 2`] = `
Array [
  Array [
    12800,
    Object {
      "id": "4abk",
      "midiEvent": Object {
        "data": Uint8Array [
          1,
          1,
          1,
        ],
        "receivedTime": 0,
      },
    },
  ],
]
`;

exports[`getEventsInRange() edge 3`] = `Array []`;
